### PR TITLE
chore: add a dns.OSConfigurator implementation that uses the CoderVPN protocol

### LINKED ISSUE #14733: Add tailscale dns.OSConfigurator that uses CoderVPN protocol
In order to support an IPv6 desktop native VPN on macOS, we need to be able to configure DNS settings to direct domain queries with the suffix `.coder` to the DNS server embedded in tailnet (#14718).  We want to do this using the native NetworkExtension APIs.  Tailscale's `wgengine` uses a `dns.OSConfigurator` to set DNS configuration.  We need to build a new implementation of this interface that uses the CoderVPN protocol (#14731) to send NetworkSettings to the manager, which then calls the native APIs.

Image

### PR DESCRIPTION
**Pull request by ethanndickson**
Closes #14733.

### CODE REVIEW

**Comment by ethanndickson** [comment]
* **#15342** <a href="https://app.graphite.dev/github/pr/coder/coder/15342?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ethanndickson and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: vpn/dns.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
+package vpn
+
+import "tailscale.com/net/dns"
+
+func NewVPNManager(t *Tunnel) dns.OSConfigurator {
```
Comment (spikecurtis) [inline vpn/dns.go:5]:
```suggestion
func NewDNSConfigurator(t *Tunnel) dns.OSConfigurator {
```

1. Go style typically eschews the package name from components. E.g. in `d := vpn.NewVPNConfigurator`, the second `VPN` doesn't give you much new information.
2. Let's stick with the Tailscale terminology and call this a "configurator" not a "manager".

File: vpn/dns.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			Servers:              servers,
+			SearchDomains:        searchDomains,
+			DomainName:           "coder",
+			MatchDomains:         matchDomains,
+			MatchDomainsNoSearch: false,
+		},
+	})
+}
+
+func (vpnDNSManager) GetBaseConfig() (dns.OSConfig, error) {
```
Comment (spikecurtis) [inline vpn/dns.go:37]:
```suggestion
func (*vpnDNSManager) GetBaseConfig() (dns.OSConfig, error) {
```

File: vpn/dns.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+func NewVPNManager(t *Tunnel) dns.OSConfigurator {
+	return &vpnDNSManager{tunnel: t}
+}
+
+type vpnDNSManager struct {
+	tunnel *Tunnel
+}
+
+func (v *vpnDNSManager) SetDNS(cfg dns.OSConfig) error {
```
Comment (spikecurtis) [inline vpn/dns.go:13]: needs some unit tests

File: vpn/dns.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	return true
+}
+
+// Close implements dns.OSConfigurator.
+func (*vpnDNSManager) Close() error {
+	// There's no cleanup that we need to initiate from within the dylib.
+	return nil
+}
+
+var _ dns.OSConfigurator = (*vpnDNSManager)(nil)
```
Comment (spikecurtis) [inline vpn/dns.go:55]: redundant given

```
func NewVPNManager(t *Tunnel) dns.OSConfigurator {
	return &vpnDNSManager{tunnel: t}
}
```

File: vpn/dns.go (hunk 1 above)
Comment (ethanndickson) [reply vpn/dns.go:5]: The `manager` usage came from Tailscale calling all their OSConfigurator implementation files `managers`, `manager_darwin`, etc. but sure enough they use `configurator` for all the public types.

File: vpn/dns.go (hunk 2 above)
Comment (ethanndickson) [reply vpn/dns.go:37]: I swear we had a linting rule for this

**Review by spikecurtis — APPROVED** [review]

**Pull request merged**